Add `--tag` option to `clever deploy`

Right now, with the new isomorphic-git based clever-tools, we cannot deploy a tag. We have identified why isomorphic-git cannot do it (for now), filed [an issue](https://github.com/isomorphic-git/isomorphic-git/issues/520) and proposed [a solution](https://github.com/isomorphic-git/isomorphic-git/pull/521).

When this is available, we'll be able to add an option to `clever deploy` called `--tag` to push and deploy a specific tag to the clever remote.

Thanks @urcadox for your reply !

So, I'm using `$CI_COMMIT_REF_NAME` Gitlab's variable who contains the name of the branch / tag. The first echo in my Gitlab CI script above display well the name of my tag:

```shell
cd ./dist

cp -r ./../.git ./

echo -e "\n\nGit tag is $CI_COMMIT_REF_NAME\n\n"

git checkout $CI_COMMIT_REF_NAME

echo "Sources:"
ls -al
echo -e "\n\n"

echo "Installing clever-tools ..."
npm i clever-tools
echo "clever-tools installed"

echo "Logging to Clever Cloud with token ..."
./node_modules/clever-tools/bin/clever.js login --token $CC_TOKEN_DEV --secret $CC_SECRET_DEV
echo "Logged"

echo "Linking to app ..."
./node_modules/clever-tools/bin/clever.js link $APP_ID_BOT_DEV

echo "Deploying ..."
./node_modules/clever-tools/bin/clever.js deploy --verbose --branch $CI_COMMIT_REF_NAME --force
```

I got this error:
```shell
[ERROR] Object 3d36ef0b98af39d9a6258b8025ffbeb3a895c31d was anticipated to be a commit but it is a tag. This is probably a bug deep in isomorphic-git!
```

--- a/bin/clever.js
+++ b/bin/clever.js
@@ -183,6 +183,12 @@ function run () {
       metavar: 'commit id',
       description: 'Restart the application with a specific commit ID',
     }),
+    gitTag: cliparse.option('tag', {
+      aliases: ['t'],
+      default: '',
+      metavar: 'tag',
+      description: 'Tag to push (none by default)',
+    }),
     databaseId: cliparse.option('database-id', {
       metavar: 'database_id',
       description: 'The Database ID (e.g.: postgresql_xxx)',
@@ -646,7 +652,7 @@ function run () {
   const deploy = lazyRequirePromiseModule('../src/commands/deploy.js');
   const deployCommand = cliparse.command('deploy', {
     description: 'Deploy an application',
-    options: [opts.alias, opts.branch, opts.quiet, opts.forceDeploy, opts.followDeployLogs, opts.sameCommitPolicy],
+    options: [opts.alias, opts.branch, opts.gitTag, opts.quiet, opts.forceDeploy, opts.followDeployLogs, opts.sameCommitPolicy],
   }, deploy('deploy'));
 
   // DIAG COMMAND

--- a/package.json
+++ b/package.json
@@ -32,7 +32,7 @@
     "common-env": "^6.4.0",
     "curlconverter": "^3.21.0",
     "eventsource": "^1.1.0",
-    "isomorphic-git": "^1.25.0",
+    "isomorphic-git": "^1.25.3",
     "linux-release-info": "^3.0.0",
     "lodash": "^4.17.21",
     "mkdirp": "^1.0.4",

--- a/src/commands/deploy.js
+++ b/src/commands/deploy.js
@@ -13,12 +13,12 @@ const { sendToApi } = require('../models/send-to-api.js');
 // Once the API call to redeploy() has been triggered successfully,
 // the rest (waiting for deployment state to evolve and displaying logs) is done with auto retry (resilient to network failures)
 async function deploy (params) {
-  const { alias, branch: branchName, quiet, force, follow, 'same-commit-policy': sameCommitPolicy } = params.options;
+  const { alias, branch: branchName, tag: tagName, quiet, force, follow, 'same-commit-policy': sameCommitPolicy } = params.options;
 
   const appData = await AppConfig.getAppDetails({ alias });
   const { ownerId, appId } = appData;
-  const branchRefspec = await git.getFullBranch(branchName);
 
+  const branchRefspec = await getBranchToDeploy(branchName, tagName);
   const commitIdToPush = await git.getBranchCommit(branchRefspec);
   const remoteHeadCommitId = await git.getRemoteCommit(appData.deployUrl);
   const deployedCommitId = await Application.get(ownerId, appId)
@@ -66,7 +66,8 @@ async function deploy (params) {
   const knownDeployments = await getAllDeployments({ id: ownerId, appId, limit: 5 }).then(sendToApi);
 
   Logger.println('Pushing source code to Clever Cloud…');
-  await git.push(appData.deployUrl, branchRefspec, force)
+
+  await git.push(appData.deployUrl, commitIdToPush, force)
     .catch(async (e) => {
       const isShallow = await git.isShallow();
       if (isShallow) {
@@ -84,6 +85,22 @@ async function deploy (params) {
 async function restartOnSameCommit (ownerId, appId, commitIdToPush, quiet, follow, withoutCache) {
   const restart = await Application.redeploy(ownerId, appId, commitIdToPush, withoutCache);
   return Log.watchDeploymentAndDisplayLogs({ ownerId, appId, deploymentId: restart.deploymentId, quiet, follow });
+}
+
+async function getBranchToDeploy (branchName, tagName) {
+  if (tagName) {
+    const useTag = await git.isExistingTag(tagName);
+    if (useTag) {
+      const tagRefspec = await git.getFullBranch(tagName);
+      return tagRefspec;
+    }
+    else {
+      throw new Error(`Tag ${tagName} doesn't exist locally`);
+    }
+  }
+  else {
+    return await git.getFullBranch(branchName);
+  }
 }
 
 module.exports = { deploy };

--- a/src/models/git.js
+++ b/src/models/git.js
@@ -78,7 +78,19 @@ async function getFullBranch (branchName) {
 
 async function getBranchCommit (refspec) {
   const repo = await getRepo();
-  return git.resolveRef({ ...repo, ref: refspec });
+  const oid = await git.resolveRef({ ...repo, ref: refspec });
+  // When a refspec refers to an annotated tag, the OID ref represents the annotation and not the commit directly,
+  // that's why we need a call to `readCommit`.
+  const res = await git.readCommit({ ...repo, ref: refspec, oid });
+  return res.oid;
+}
+
+async function isExistingTag (tag) {
+  const repo = await getRepo();
+  const tags = await git.listTags({
+    ...repo,
+  });
+  return tags.includes(tag);
 }
 
 async function push (remoteUrl, branchRefspec, force) {
@@ -131,4 +143,5 @@ module.exports = {
   push,
   completeBranches,
   isShallow,
+  isExistingTag,
 };

